x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/extractor/ExtractedFieldsDetector.java

```diff
@@ -86,6 +88,10 @@ public ExtractedFields detect() {
         if (extractedFields.getAllFields().isEmpty()) {
             throw ExceptionsHelper.badRequestException("No compatible fields could be detected in index [{}]", index);
         }
+        if (extractedFields.getAllFields().size() > docValueFieldsLimit) {
+            extractedFields = new ExtractedFields(extractedFields.getAllFields().stream().map(ExtractedField::newFromSource)
+                .collect(Collectors.toList()));
```

Comment:
It seems to me that we only really need to transform as many as necessary to drop below the `docValueFieldsLimit`, and then only those that that `supportsFromSource()`.

Reply:
>we only really need to transform as many as necessary to drop below the docValueFieldsLimit

As long as we have to touch the source, we might as well fetch them all from there. Performance-wise it's going to be better.

>and then only those that that supportsFromSource()

But this is a very good poing. In case we have fields that don't support from-source, we should insist taking them from doc_values. Then we'll also need a check to see if we're still over the limit.